dask/array/core.py

```diff
@@ -629,11 +629,19 @@ def map_blocks(func, *args, **kwargs):
             vv = v
             v = v[0]
             [(key, task)] = v.dsk.items()  # unpack subgraph callable
+
+            if new_axis:
+                # anything using the keys in dsk is incorrect, as the
```

Comment:
Yeah, unfortunately map_blocks has become a bit choked with features so that it's getting pretty gnarly in here and hard to tell what's going on with confidence.